api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirEncounterDaoImpl.java

```diff
@@ -44,4 +57,116 @@ public Encounter getEncounterByUuid(String uuid) {
 		        .createAlias("patient", "p").createAlias("p.identifiers", "pi").add(eq("pi.identifier", patientIdentifier))
 		        .list();
 	}
+	
+	@Override
+	public Collection<Encounter> searchForEncounters(DateRangeParam date, ReferenceParam location,
+	        ReferenceParam participant, ReferenceParam subject, TokenOrListParam participantIdentifier,
+	        StringOrListParam participantName, StringOrListParam participantGiven, StringOrListParam participantFamily,
+	        TokenOrListParam subjectIdentifier, StringOrListParam subjectName, StringOrListParam subjectGiven,
+	        StringOrListParam subjectFamily) {
+		
+		Criteria criteria = sessionFactory.getCurrentSession().createCriteria(Encounter.class);
+		handleSubjectNames(criteria, subjectName, subjectGiven, subjectFamily);
+		handleParticipantNames(criteria, participantName, participantGiven, participantFamily);
+		handleSubjectIdentifier(criteria, subjectIdentifier);
+		handleParticipantIdentifier(criteria, participantIdentifier);
+		handleDateRange("date", date).ifPresent(criteria::add);
+		//TODO: handle ReferenceParam for participant, subject, location
+		
+		return criteria.list();
+	}
+	
+	private void handleParticipantIdentifier(Criteria criteria, TokenOrListParam identifier) {
+		if (identifier == null) {
+			return;
+		}
+		
+		criteria.createAlias("identifiers", "pi", JoinType.INNER_JOIN, eq("pi.voided", false));
+		
+		handleOrListParamBySystem(identifier, (system, tokens) -> {
+			if (system.isEmpty()) {
+				return Optional.of(in("pi.identifier", tokensToList(tokens)));
+			} else {
+				if (!containsAlias(criteria, "pit")) {
+					criteria.createAlias("pi.identifierType", "pit");
+				}
+				
+				return Optional.of(and(eq("pit.name", system), in("pi.identifier", tokensToList(tokens))));
+			}
+		}).ifPresent(criteria::add);
+	}
+	
+	private void handleSubjectIdentifier(Criteria criteria, TokenOrListParam identifier) {
+		if (identifier == null) {
+			return;
+		}
+		
+		criteria.createAlias("identifiers", "si", JoinType.INNER_JOIN, eq("si.voided", false));
+		
+		handleOrListParamBySystem(identifier, (system, tokens) -> {
+			if (system.isEmpty()) {
+				return Optional.of(in("si.identifier", tokensToList(tokens)));
+			} else {
+				if (!containsAlias(criteria, "sit")) {
+					criteria.createAlias("si.identifierType", "sit");
+				}
+				
+				return Optional.of(and(eq("sit.name", system), in("si.identifier", tokensToList(tokens))));
+			}
+		}).ifPresent(criteria::add);
+	}
+	
+	private void handleSubjectNames(Criteria criteria, StringOrListParam name, StringOrListParam given,
```

Comment:
I noticed that a similar handler is used for `PatientDaoImpl` and will be used for `PersonDaoImpl` as well, shouldn't we refactor this to be present in the `BaseDaoImpl`?

Reply:
Yes, we should